sycl/include/CL/sycl/atomic_ref.hpp

```diff
@@ -0,0 +1,675 @@
+//==----- atomic_ref.hpp - SYCL 2020 atomic_ref ----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/__spirv/spirv_ops.hpp>
+#include <CL/sycl/access/access.hpp>
+#include <CL/sycl/atomic.hpp>
+#include <CL/sycl/detail/defines.hpp>
+#include <CL/sycl/detail/spirv.hpp>
+#include <CL/sycl/memory_enums.hpp>
+#ifdef __SYCL_DEVICE_ONLY__
+#include <CL/sycl/multi_ptr.hpp>
+#endif
+
+#ifndef __SYCL_DEVICE_ONLY__
+#include <atomic>
+#endif
+#include <type_traits>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+using memory_order = cl::sycl::memory_order;
+using memory_scope = cl::sycl::memory_scope;
+
+#if __cplusplus < 201703L
```

Comment:
This code is required for C++ < 17 (C++14 in our case) since `memory_enums.hpp` defines such constants for C++17 only. This is not a protection from `inline constexpr`.

Reply:
Could there be a benefit to removing `if __cplusplus >= 201703L` in both this and memory_enums.hpp, then replace `inline constexpr` in memory_enums.hpp with `__SYCL_INLINE_CONSTEXPR`?

Reply:
It's for consistency with ISO C++.  C++11 had an unscoped enum, and C++20 swapped it for a scoped enum.  My understanding is that the `inline constexpr` versions were introduced for backwards compatibility with the old names. 
 See https://en.cppreference.com/w/cpp/atomic/memory_order.

We could have said that SYCL 2020 had an unscoped enum (for consistency with C++17) and then had SYCL-Next define `inline constexpr` versions for backward compatibility (for consistency with C++20), but since `inline constexpr` works in C++17 there didn't seem to be any harm in aligning with the C++20 version right away.

Reply:
@steffenlarsen @gmlueck @Pennycook As I understand, it's better to keep things as they are today?

Reply:
@gmlueck This code is protected from C++17, not from C++14 (**<** 201703L), but the code is required only because the constants is used inside the implementation of `sycl::atomic_ref`. I think it is better to remove this code and use `memory_order::relaxed` instead of `memory_order_relaxed`, etc in the class.

Reply:
Oh, I read the code too quickly.  I'm confused about the macro test now.  Why isn't the test `#if __cplusplus >= 201703L`?

Yes, it makes sense to use `memory_order::relaxed` instead of `memory_order_relaxed` in the implementation of the header.  The SYCL spec still requires the `memory_order_relaxed` variables to be defined, though.

Reply:
@gmlueck This code added the constants for C++14 only since for C++17 (and newer) they are added by `sycl/memory_enums.hpp`. I've removed this code. Also, could you have a look at my latest question in the PR (about `generic_space`)?